ui/src/chat/ChatMessage/ChatMessage.tsx

```diff
@@ -57,19 +58,45 @@ const ChatMessage = React.memo<
       ref
     ) => {
       const { seal, memo } = writ;
-      const { markRead } = useChatState.getState();
-      const [viewRef, inView, entry] = useInView({
+      const chatInfo = useChatInfo(whom);
+      const unread = chatInfo?.unread;
+      const { ref: viewRef } = useInView({
         threshold: 1,
-        triggerOnce: true,
+        onChange: useCallback(
```

Comment:
Optional: Since this is a long function with stacked conditionals, would improve readability to have a comment to explain the decision tree.  Not a merge-blocker.